pkg/services/internal/whatsonchain/utils.go

```diff
@@ -0,0 +1,81 @@
+package whatsonchain
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/wdk"
+	"github.com/bsv-blockchain/go-sdk/transaction"
+)
+
+func extractRawTransactions(beef *transaction.Beef, txIDs []string) ([][]byte, error) {
+	rawTxs := make([][]byte, len(txIDs))
+	for i, txid := range txIDs {
+		tx := beef.FindTransaction(txid)
+		if tx == nil {
+			return nil, fmt.Errorf("cannot find transaction %s in BEEF", txid)
+		}
+		raw := tx.Bytes()
+		if len(raw) == 0 {
+			return nil, fmt.Errorf("empty raw transaction for %s", txid)
+		}
+		rawTxs[i] = raw
+	}
+	return rawTxs, nil
+}
+
+func waitOrCancel(ctx context.Context, delay time.Duration, txid string) error {
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("context canceled while waiting for tx %s: %w", txid, ctx.Err())
+	case <-time.After(delay):
+		return nil
+	}
+}
+
+func classifyBroadcastStatus(status BroadcastStatus) (wdk.PostedTxIDResultStatus, []string) {
+	switch status {
+	case StatusSuccess:
+		return wdk.PostedTxIDResultSuccess, nil
+	case StatusAlreadyBroadcasted:
+		return wdk.PostedTxIDResultAlreadyKnown, []string{"Transaction already in mempool"}
+	case StatusDoubleSpend:
+		return wdk.PostedTxIDResultDoubleSpend, []string{"Double spend detected"}
+	case StatusMissingInputs:
+		return wdk.PostedTxIDResultMissingInputs, []string{"Missing inputs detected"}
+	case StatusError:
+		return wdk.PostedTxIDResultError, []string{"Broadcast status error"}
+	default:
+		return wdk.PostedTxIDResultError, []string{"Unknown error"}
+	}
+}
+
+func convertNotes(notes []string) wdk.Notes {
+	converted := make(wdk.Notes, len(notes))
+	for i, note := range notes {
+		now := time.Now()
+		converted[i] = wdk.ReqHistoryNote{
+			When: &now,
+			What: note,
+		}
+	}
+	return converted
+}
+
+func containsI(a string, b string) bool {
+	return strings.Contains(
+		strings.ToLower(a),
+		strings.ToLower(b),
+	)
+}
```

Comment:
Consider checking a bunch of strings in a single call:
```suggestion
func containsI(subject string, contains ...string) bool {
	subject = strings.ToLower(subject)

	for _, c := range contains {
		if strings.Contains(subject, strings.ToLower(c)) {
			return true
		}
	}

	return false
}
```

This way, you could simplify checking for multiple strings:
```golang
switch {
		case containsI(responseText, "already in mempool", "txn-already-known"):
```